src/pcap/cmd/pcap-bosh-cli/main.go

```diff
@@ -0,0 +1,472 @@
+	return nil
+}
+
+// Config represents the content of the bosh config-file (default location: ~/.bosh/config).
+//
+// It contains a list of bosh Environments.
+type Config struct {
+	Environments []Environment `yaml:"environments"`
+}
+
+// Environment contains all the necessary information to connect to a specific bosh-director.
+type Environment struct {
+	AccessToken     string       `yaml:"access_token" validate:"required"`
+	AccessTokenType string       `yaml:"access_token_type" validate:"required"`
+	Alias           string       `yaml:"alias" validate:"required"`
+	CaCert          string       `yaml:"ca_cert" validate:"required"`
+	RefreshToken    string       `yaml:"refresh_token" validate:"required"`
+	URL             string       `yaml:"url" validate:"required,url"`
+	DirectorURL     *url.URL     `yaml:"-"`
+	UaaURL          *url.URL     `yaml:"-"`
+	client          *http.Client `yaml:"-"`
+}
+
+// UpdateTokens is the public wrapper func for the bosh Environment struct
+//
+// It fetches the bosh-uaa URLs from the bosh-director (if necessary)
+// and refreshes the bosh authentication tokens.
+func (e *Environment) UpdateTokens() error {
+	if e.UaaURL == nil {
+		err := e.init()
+		if err != nil {
+			return err
+		}
+		err = e.fetchUAAURL()
+		if err != nil {
+			return err
+		}
+	}
+	err := e.refreshTokens()
+	if err != nil {
+		return fmt.Errorf("failed to refresh bosh access token %w", err)
+	}
+	return nil
+}
+
+// init sets up a Bosh environment by parsing the bosh-director URL (string) from the config-file
+// and then sets up the http client for use with either TLS or plain HTTP.
+func (e *Environment) init() error {
+	var err error
+	e.DirectorURL, err = url.Parse(e.URL)
+	if err != nil {
+		return fmt.Errorf("error parsing environment url (%v) %w", e.URL, err)
+	}
+
+	if e.DirectorURL.Scheme == "https" {
+		logger.Info("using TLS-encrypted connection to bosh-director", zap.String("bosh-director-url", e.DirectorURL.String()))
+		boshCA := x509.NewCertPool()
+		ok := boshCA.AppendCertsFromPEM([]byte(e.CaCert))
+		if !ok {
+			return fmt.Errorf("could not add BOSH Director CA from bosh-config, adding to the cert pool failed %v", e.CaCert) // TODO really output cert here?
+		}
+
+		transport := http.DefaultTransport.(*http.Transport).Clone()
+		transport.TLSClientConfig.RootCAs = boshCA
+
+		e.client = &http.Client{
+			Transport: transport,
+		}
+	} else {
+		logger.Warn("using unencrypted connection to bosh-director", zap.String("bosh-director-url", e.DirectorURL.String()))
+		e.client = http.DefaultClient
+	}
+	return nil
+}
+
+// fetchUAAURL connects to the bosh-director API to fetch the bosh-uaa API URL.
+func (e *Environment) fetchUAAURL() error {
+	res, err := e.client.Do(&http.Request{ //nolint:bodyclose // closed via pcap.CloseQuietly below.
+		Method: http.MethodGet,
+		URL: &url.URL{
+			Scheme: e.DirectorURL.Scheme,
+			Host:   e.DirectorURL.Host,
+			Path:   "/info",
```

Comment:
Consider using `JoinPath` if that works (see other comment)

Reply:
Done.